pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java

```diff
@@ -321,17 +326,296 @@ private static void setValuesToMap(String keyColumnName, String valueColumnName,
+   * Examples:
+   * - jsonExtractKey('{"a": 1, "b": {"c": 2}}', '$.*', 1) returns ["$['a']", "$['b']"]
+   * - jsonExtractKey('{"a": 1, "b": {"c": 2}}', '$.*', 2) returns ["$['a']", "$['b']"]
+   * - jsonExtractKey('{"a": [1, 2]}', '$.*', 1) returns ["$['a']"]
+   *
+   * @param jsonObj  JSON object or string
+   * @param jsonPath JsonPath expression to extract keys
+   * @param maxDepth Maximum depth to recurse (must be positive)
+   * @return List of key paths matching the JsonPath up to maxDepth, empty list if input is null or invalid
+   */
+  @ScalarFunction
+  public static List jsonExtractKey(Object jsonObj, String jsonPath, int maxDepth)
+      throws IOException {
+    return jsonExtractKey(jsonObj, jsonPath, maxDepth, false);
+  }
+
+  /**
+   * Extract keys from JSON object using JsonPath with depth limit and output format option.
+   * <p>
+   * Examples:
+   * - jsonExtractKey('{"a": 1, "b": {"c": 2}}', '$.*', 1, false) returns ["$['a']", "$['b']"]
+   * - jsonExtractKey('{"a": 1, "b": {"c": 2}}', '$..**', 2, true) returns ["a", "b", "b.c"]
+   * - jsonExtractKey('{"a": [1, 2]}', '$.*', 1, true) returns ["a"]
+   *
+   * @param jsonObj     JSON object or string
+   * @param jsonPath    JsonPath expression to extract keys
+   * @param maxDepth    Maximum depth to recurse (must be positive)
+   * @param dotNotation If true, return keys in dot notation (e.g., "a.b.c"),
+   *                    if false, return JsonPath format (e.g., "$['a']['b']['c']")
+   * @return List of key paths matching the JsonPath up to maxDepth, empty list if input is null or invalid
+   */
+  @ScalarFunction
+  public static List jsonExtractKey(Object jsonObj, String jsonPath, int maxDepth, boolean dotNotation)
+      throws IOException {
+    if (maxDepth <= 0) {
+      return java.util.Collections.emptyList();
+    }
+
+    // Special handling for $.** and $.. recursive key extraction
+    if ("$..**".equals(jsonPath) || "$..".equals(jsonPath)) {
+      return extractAllKeysRecursively(jsonObj, maxDepth, dotNotation);
+    }
+
+    // For other expressions, try to get keys using AS_PATH_LIST
+    List<String> keys = null;
+    try {
+      keys = KEY_PARSE_CONTEXT.parse(
+          jsonObj instanceof String ? (String) jsonObj : jsonObj).read(jsonPath);
+    } catch (Exception e) {
+      // AS_PATH_LIST might not work for all expressions
+    }
+
+    // If AS_PATH_LIST doesn't work, fall back to manual path construction
+    if (keys == null || keys.isEmpty()) {
+      return extractKeysForNonRecursiveExpression(jsonObj, jsonPath, maxDepth, dotNotation);
+    }
+
+    // Filter keys by depth if maxDepth is specified
+    if (maxDepth != Integer.MAX_VALUE) {
+      keys = keys.stream()
+          .filter(key -> getKeyDepth(key) <= maxDepth)
+          .collect(java.util.stream.Collectors.toList());
+    }
+
+    // Convert to dot notation if requested
+    if (dotNotation) {
+      keys = keys.stream()
+          .map(JsonFunctions::convertToDotNotation)
```

Comment:
Are we converting it twice? Seems it is already converted in `extractKeysFromNode()`